Add Per-assembly Load Native Library callbacks

swaroop-sridhar opened the pull request:
This Change implements the Native Library resolution
Call-backs proposed in dotnet/corefx#32015

pentp commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -32,6 +32,11 @@ public static partial class Marshal
         internal static Guid IID_IUnknown = new Guid("00000000-0000-0000-C000-000000000046");
 #endif //FEATURE_COMINTEROP
 
+        static Marshal()
```
An explicit static constructor is bad for performance and in this case also causes an infinite loop during startup. Maybe `nativeDllResolveMap` should be lazy-initialized on first use?

jkotas replied:
Agree

jkotas commented:
> This API is not yet approved, and the API contracts in CoreFX will not be added until the API approval is complete

That's not the usual process...

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1907,13 +1913,87 @@ public static bool TryGetLibraryExport(IntPtr handle, string name, out IntPtr ad
+        /// Therefore, this table uses weak assembly pointers to indirectly achieve 
+        /// similar behavior.
+        /// </summary>
+        public static ConditionalWeakTable<Assembly, Func<string, Assembly, DllImportSearchPath?, IntPtr>> nativeDllResolveMap;
```
From https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md: prefix static fields with `s_`

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1907,13 +1913,87 @@ public static bool TryGetLibraryExport(IntPtr handle, string name, out IntPtr ad
+        internal static IntPtr LoadLibraryCallbackStub(string libraryName, Assembly assembly, 
+                                                       bool hasDllImportSearchPathFlags, uint dllImportSearchPathFlags)
+        {
+            Func<string, Assembly, DllImportSearchPath?, IntPtr> callBack;
```
`callback` is one word. `B` should be lowercase.

jkotas replied:
(multiple places)

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6280,23 +6280,32 @@ INT_PTR NDirect::GetNativeLibraryExport(NATIVE_LIBRARY_HANDLE handle, LPCWSTR sy
 }
 
+#ifndef PLATFORM_UNIX
+BOOL IsWindowsAPI(PCWSTR wszLibName)
```
This does not return whether the name is Windows API. It returns whether the name is Windows API set. Change the name to `IsWindowsAPISet` ?

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6362,6 +6371,55 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaHost(NDirectMethodDesc * pMD,
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaCallBack(NDirectMethodDesc * pMD, LPCWSTR wszLibName)
+{
+#ifndef PLATFORM_UNIX
+    if (IsWindowsAPI(wszLibName))
```
What is the rationale behind blocking this?

AaronRobinsonMSFT replied:
My only concern with not blocking this is the scenario where we deploy Windows API sets on downlevel Windows SKUs  (i.e. Windows 7). The issue here would be why are we looking in the system if they are app local? This is largely a hedge about not trusting Windows to deploy API sets that are guaranteed backwards compatible. I think that is a fair assumption the API sets will always be backwards compat, but note that the behavior will be different on different platforms if we rely on a system API set in some cases and our applocal version of the API sets in others.

jkotas commented on src/vm/callhelpers.h:
```diff
@@ -566,6 +566,7 @@ enum DispatchCallSimpleFlags
 #define STRINGREF_TO_ARGHOLDER(x) (LPVOID)STRINGREFToObject(x)
 #define PTR_TO_ARGHOLDER(x) (LPVOID)x
 #define DWORD_TO_ARGHOLDER(x)   (LPVOID)(SIZE_T)x
+#define BOOL_TO_ARGHOLDER(x) DWORD_TO_ARGHOLDER(x)   
```
This is not quite right. Assuming the managed argument is going to be `bool`, this value should be normalized, e.g. using `!!`

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6362,6 +6371,55 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaHost(NDirectMethodDesc * pMD,
+
+    Assembly* pAssembly = pMD->GetMethodTable()->GetAssembly();
+    protect.libNameRef = StringObject::NewString(wszLibName);
+    protect.assemblyRef = pAssembly->GetExposedObject();
```
This has GC hole. You are assigning object references without them being protected.

This exact situation is described in "Your first GC hole" in https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.1.2.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6362,6 +6371,55 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaHost(NDirectMethodDesc * pMD,
 }
 
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaCallBack(NDirectMethodDesc * pMD, LPCWSTR wszLibName)
+{
```
Add contracts to all new methods.

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1907,13 +1913,87 @@ public static bool TryGetLibraryExport(IntPtr handle, string name, out IntPtr ad
+                return false;
+            }
+
+            nativeDllResolveMap.Add(assembly, callBack);
```
This has race condition: if the method is called from multiple threads on same assembly, we may end up throwing here.

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs:
```diff
@@ -1907,13 +1913,87 @@ public static bool TryGetLibraryExport(IntPtr handle, string name, out IntPtr ad
+        /// and returns
+        ///   - The handle to the loaded native library (on success) or null (on failure)
+        /// The parameters on this callback are such that they can be directly passed to 
+        /// Marhall.LoadLibrary(libraryName, assembly, dllImportSearchPath) to approximately achieve 
```
```suggestion
        /// Marhal.LoadLibrary(libraryName, assembly, dllImportSearchPath) to approximately achieve 
```

AaronRobinsonMSFT replied:
`Marshal` ?

AaronRobinsonMSFT commented on tests/src/Interop/MarshalAPI/DllImportUtil/DllImportUtil.cs:
```diff
@@ -0,0 +1,110 @@
+
+namespace DllImportUtil
+{
+    public enum TestResult {
```
I really don't see why these error codes are needed. It almost seems like an entire sub-test framework is being built here. Tests should just throw an exception with a message. All this infrastructure just makes the testing appear more complex than it needed. We also already have the `Assert` API that should be used instead of custom solutions that bifurcate having consistent tests.

swaroop-sridhar commented:
@jkotas @AaronRobinsonMSFT @pentp I have addressed your feedback in the new commit.
`NativeLibrary` APIs are approved. The API now lives in the `NativeLibrary` class instead of `Marshal`

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +178,74 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+
+            if (s_nativeDllResolveMap == null)
+            {
+                s_nativeDllResolveMap = new ConditionalWeakTable<Assembly, DllImportResolver>();
```
This needs to use Interlock.CompareExchange

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +178,74 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+            catch (ArgumentException e)
+            {
+                // ConditionalWealTable throws ArgumentException if the Key already exists
+                throw new InvalidOperationException("Resolver is alerady Set for the Assembly");
```
This needs to be localizable string

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +178,74 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+        {
+            DllImportResolver resolver;
+
+            if (!s_nativeDllResolveMap.TryGetValue(assembly, out resolver))
```
This needs to handle `s_nativeDllResolveMap == null`.

This also suggests that you may have a hole in the test coverage for this change if there are no tests failing because of this.

jkotas replied:
```suggestion
            if (!s_nativeDllResolveMap.TryGetValue(assembly, out  DllImportResolver resolver))
```

pentp commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +179,80 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+
+            if (s_nativeDllResolveMap == null)
+            {
+                Interlocked.Exchange(ref s_nativeDllResolveMap,
```
This should be `Interlocked.CompareExchange(ref s_nativeDllResolveMap, new ConditionalWeakTable<Assembly, DllImportResolver>(), null);` so that it's guaranteed to be initialized only once.

swaroop-sridhar commented:
@dotnet-bot test Windows_NT x64 Release CoreFX Tests 
@dotnet-bot  test OSX10.12 x64 Checked Innerloop Build and Test

swaroop-sridhar commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +179,78 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+        /// Therefore, this table uses weak assembly pointers to indirectly achieve 
+        /// similar behavior.
+        /// </summary>
+        public static ConditionalWeakTable<Assembly, DllImportResolver> s_nativeDllResolveMap = null;
```
Fixed it, thanks

jkotas commented on src/System.Private.CoreLib/Resources/Strings.resx:
```diff
@@ -2428,7 +2428,10 @@
   </data>
-  <data name="InvalidOperation_CannotRemoveLastFromEmptyCollection" xml:space="preserve">
+  <data name="InvalidOperation_CannotRegisterSecondResolver" xml:space="preserve">
+    <value>A Resolver is alerady Set for the Assembly.</value>
```
Type: alerady

jkotas replied:
Typo: Set in the middle of the sentence should be lower case.

swaroop-sridhar replied:
Fixed it, thanks.

jkotas commented:
Looks good to me, but it can use a second pair of eyes. I have looked at this too many times...

AaronRobinsonMSFT commented:
@jkotas I will take another look.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -9,9 +9,26 @@
 
 namespace System.Runtime.InteropServices
 {
+
```
Remove extra line.

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -9,9 +9,26 @@
+    /// </summary>
+    /// <param name="libraryName">The native library to resolve</param>
+    /// <param name="assembly">The assembly requesting the resolution</param>
+    /// <param name="DllImportSearchPath?">
```
The name is actually `searchPath` - you have the type 😄

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -161,6 +179,78 @@ public static bool TryGetExport(IntPtr handle, string name, out IntPtr address)
+            {
+                s_nativeDllResolveMap.Add(assembly, resolver);
+            }
+            catch (ArgumentException e)
```
If we are not going to use the exception, then I would remove it (i.e. remove the `e`). Alternatively we could add it as the inner exception of the thrown `InvalidOperationException`?

AaronRobinsonMSFT commented on src/vm/dllimport.h:
```diff
@@ -122,12 +122,13 @@ class NDirect
     static NATIVE_LIBRARY_HANDLE LoadLibraryModuleViaHost(NDirectMethodDesc * pMD, LPCWSTR wszLibName);
     static NATIVE_LIBRARY_HANDLE LoadLibraryModuleViaEvent(NDirectMethodDesc * pMD, LPCWSTR wszLibName);
+    static NATIVE_LIBRARY_HANDLE LoadLibraryModuleViaCallback(NDirectMethodDesc * pMD, LPCWSTR wszLibName);
     static NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(NDirectMethodDesc * pMD, LoadLibErrorTracker * pErrorTracker, const wchar_t* wszLibName);
```
I realize this isn't in your review, but having a raw `wchar_t` should be avoided.

AaronRobinsonMSFT commented on src/vm/dllimport.h:
```diff
@@ -122,12 +122,13 @@ class NDirect
+    static NATIVE_LIBRARY_HANDLE LoadLibraryModuleViaCallback(NDirectMethodDesc * pMD, LPCWSTR wszLibName);
     static NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(NDirectMethodDesc * pMD, LoadLibErrorTracker * pErrorTracker, const wchar_t* wszLibName);
-    static NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(Assembly *callingAssembly, BOOL searchAssemblyDirectory, DWORD dllImportSearchPathFlag, LoadLibErrorTracker * pErrorTracker, const wchar_t* wszLibName);
+    static NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(Assembly *callingAssembly, BOOL searchAssemblyDirectory, DWORD dllImportSearchPathFlags, LoadLibErrorTracker * pErrorTracker, const wchar_t* wszLibName);
```
Use `LPCWSTR` instead of `const wchar_t *`.

AaronRobinsonMSFT commented on src/vm/interoputil.h:
```diff
@@ -137,7 +137,7 @@ void FillExceptionData(
 //returns true if pImport has DefaultDllImportSearchPathsAttribute
-//if true, also returns dllImportSearchPathFlag and searchAssemblyDirectory values.
+//if true, also returns dllImportSearchPathFlags and searchAssemblyDirectory values.
 BOOL GetDefaultDllImportSearchPathsAttributeValue(IMDInternalImport *pImport, mdToken token, DWORD * pDlImportSearchPathFlag);
```
Based on these changes should `pDlImportSearchPathFlag` be renamed to `pDlImportSearchPathFlags`?

AaronRobinsonMSFT commented on tests/issues.targets:
```diff
@@ -391,6 +391,9 @@
         <ExcludeList Include="$(XunitTestBinBase)/Interop/NativeLibraryResolveEvent/ResolveEventTests/*">
             <Issue>21964</Issue>
         </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/Interop/NativeLibraryResolveCallBack/CallBackTests/*">
```
I know this is a huge pain, but we should make this consistent. Please update these paths so they are cased correctly - `NativeLibraryResolveCallback` (i.e. callback is one word).

AaronRobinsonMSFT replied:
Throughout.

swaroop-sridhar replied:
I'll update the name of this test. But there are a lot of places in the source code, and many tests that use the CallBack notation. I don't want to touch all that in this checkin (~2000 changes), can do it separately.

AaronRobinsonMSFT replied:
@swaroop-sridhar Please don't take this as do it everywhere in the repo. Definitely don't want you to do that. I just don't want to add any more.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6203,29 +6245,13 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleBySearch(NDirectMethodDesc * pMD
-            searchAssemblyDirectory = pModule->DllImportSearchAssemblyDirectory();
-        }
-    }
+    GetDllImportSearchPathFlags(pMD, &dllImportSearchPathFlags, &searchAssemblyDirectory);
```
Check this return value.

swaroop-sridhar replied:
Why should we check this value? We use the values in by-ref parameters either way. 
TRUE case: Attribute was found, by-ref parameters hold that value.
FALSE case: Attribute was not found. by-ref parameters hold our default chosen values.
This is true for calls at 6227 and 6251.

AaronRobinsonMSFT replied:
Yes. I misread that function.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6127,6 +6122,57 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    {
+        *dllImportSearchPathFlags = pModule->DefaultDllImportSearchPathsAttributeCachedValue();
+        *searchAssemblyDirectory = pModule->DllImportSearchAssemblyDirectory();
+        return true;
```
`return TRUE;`

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6127,6 +6122,57 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    }
+
+    GetDefaultDllImportSearchPathFlags(dllImportSearchPathFlags, searchAssemblyDirectory);
+    return false;
```
`return FALSE;`

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6127,6 +6122,57 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+    {
+        *dllImportSearchPathFlags = pMD->DefaultDllImportSearchPathsAttributeCachedValue();
+        *searchAssemblyDirectory = pMD->DllImportSearchAssemblyDirectory();
+        return true;
```
`return TRUE;`

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6164,30 +6210,26 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryByName(LPCWSTR libraryName, Assembly *
-            dllImportSearchPathFlag = pModule->DefaultDllImportSearchPathsAttributeCachedValue();
-            searchAssemblyDirectory = pModule->DllImportSearchAssemblyDirectory();
-        }
+        GetDllImportSearchPathFlags(callingAssembly->GetManifestModule(),
```
Check return value.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6438,6 +6474,51 @@ NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryModuleViaEvent(NDirectMethodDesc * pMD
+
+    DWORD dllImportSearchPathFlags;
+    BOOL searchAssemblyDirectory;
+    BOOL hasDllImportSearchPathFlags = GetDllImportSearchPathFlags(pMD, &dllImportSearchPathFlags, &searchAssemblyDirectory);
```
If this is false, then `dllImportSearchPathFlags` and `searchAssemblyDirectory` are in an undefined state. Please handle this case.

swaroop-sridhar replied:
If `GetDllImportSearchPathFlags` returns false, `dllImportSearchPathFlags` and `searchAssemblyDirectory` are initialized to default values.

But the contract with managed code is that `dllImportSearchPathFlags` is meaningless if `hasDllImportSearchPathFlags` is false. `LoadLibraryCallbackStub` will pass `null` to the resolver.

AaronRobinsonMSFT replied:
Yeah I am not sure what was happening here. I must have misread the code. Ignore.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6734,6 +6823,7 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
        return hmod.Extract();
     }
 
+
```
Remove extra line.

AaronRobinsonMSFT commented on src/vm/interoputil.cpp:
```diff
@@ -898,8 +898,8 @@ void FillExceptionData(
 //returns true if pImport has DefaultDllImportSearchPathsAttribute
-//if true, also returns dllImportSearchPathFlag and searchAssemblyDirectory values.
-BOOL GetDefaultDllImportSearchPathsAttributeValue(IMDInternalImport *pImport, mdToken token, DWORD * pDllImportSearchPathFlag)
+//if true, also returns dllImportSearchPathFlags and searchAssemblyDirectory values.
```
This 'and searchAssemblyDirectory' phrase should be removed. It is very confusing. The custom attribute value is being returned "unmodified" the comment alludes to some other value that is protected against elsewhere.

swaroop-sridhar commented:
Thanks for the feedback @AaronRobinsonMSFT.
I've addressed it in a306c61d0f5873fa1cfd5d3f5deaa3aa4be69ae7 -- made the changes you've suggested, except as noted in the comments.
